staging/src/k8s.io/apiserver/pkg/endpoints/filters/constrained_impersonation.go

```diff
@@ -0,0 +1,292 @@
+		// if they are specified, then they are the authority (including the inclusion of system:authenticated/system:unauthenticated groups)
+		groupsSpecified := len(req.Header[authenticationv1.ImpersonateGroupHeader]) > 0
+
+		// Building attributes to authorize, username and group information
+		var actingAsAttrsList []*authorizer.AttributesRecord
+		username := ""
+		groups := []string{}
+		userExtra := map[string][]string{}
+		uid := ""
+		for _, impersonationRequest := range impersonationRequests {
+			gvk := impersonationRequest.GetObjectKind().GroupVersionKind()
+			actingAsAttributes := &authorizer.AttributesRecord{
+				User:            attrs.GetUser(),
+				APIGroup:        gvk.Group,
+				APIVersion:      gvk.Version,
+				Namespace:       impersonationRequest.Namespace,
+				Name:            impersonationRequest.Name,
+				ResourceRequest: true,
+			}
+
+			switch gvk.GroupKind() {
+			case v1.SchemeGroupVersion.WithKind("ServiceAccount").GroupKind():
+				actingAsAttributes.Resource = "serviceaccounts"
+				actingAsAttributes.Verb = "impersonate:serviceaccount"
+				username = serviceaccount.MakeUsername(impersonationRequest.Namespace, impersonationRequest.Name)
+				if !groupsSpecified {
+					// if groups aren't specified for a service account, we know the groups because its a fixed mapping.  Add them
+					groups = serviceaccount.MakeGroupNames(impersonationRequest.Namespace)
+				}
+
+			case v1.SchemeGroupVersion.WithKind("User").GroupKind():
+				actingAsAttributes.Resource = "users"
+				actingAsAttributes.Verb = "impersonate:user-info"
+				username = impersonationRequest.Name
+
+			case v1.SchemeGroupVersion.WithKind("Group").GroupKind():
+				actingAsAttributes.Resource = "groups"
+				actingAsAttributes.Verb = "impersonate:user-info"
+				groups = append(groups, impersonationRequest.Name)
+
+			case authenticationv1.SchemeGroupVersion.WithKind("UserExtra").GroupKind():
+				extraKey := impersonationRequest.FieldPath
+				extraValue := impersonationRequest.Name
+				actingAsAttributes.Resource = "userextras"
+				actingAsAttributes.Verb = "impersonate:user-info"
+				actingAsAttributes.Subresource = extraKey
+				userExtra[extraKey] = append(userExtra[extraKey], extraValue)
+
+			case authenticationv1.SchemeGroupVersion.WithKind("UID").GroupKind():
+				uid = impersonationRequest.Name
+				actingAsAttributes.Resource = "uids"
+				actingAsAttributes.Verb = "impersonate:user-info"
+
+			default:
+				klog.V(4).InfoS("unknown impersonation request type", "request", impersonationRequest)
+				responsewriters.Forbidden(ctx, actingAsAttributes, w, req, fmt.Sprintf("unknown impersonation request type: %v", impersonationRequest), s)
+				return
+			}
+
+			actingAsAttrsList = append(actingAsAttrsList, actingAsAttributes)
+		}
+
+		decision, reason, actingAsAttributes, err := authorizeConstrainedImpersonation(ctx, requestorAttrs, actingAsAttrsList, a)
```

Comment:
I would expect this to take `groupsSpecified` as input since that is relevant to what types of constrained impersonation are allowed at all.